File: apps/api/app/schemas/system.py
"""Non-sensitive system configuration and availability contract."""

from typing import Literal

from app.schemas.common import SchemaModel


class SystemStatusRead(SchemaModel):
    environment: str
    database: Literal["connected", "unavailable"]
    upload_directory: Literal["configured", "unavailable"]
    vector_provider: Literal["Pinecone"] = "Pinecone"
    vector_status: Literal["connected", "unavailable"]
    vector_index: str
    vector_dimension: int | None
    vector_metric: str | None
    vector_count: int | None
    embedding_provider: Literal["Ollama"] = "Ollama"
    embedding_model: str
    embedding_status: Literal["connected", "unavailable"]
    embedding_dimension: int | None
    compatibility: Literal["compatible", "dimension_mismatch", "unavailable"]
    llm_provider: Literal["Ollama"] = "Ollama"
    llm_model: str
    llm: Literal["connected", "unavailable"]
    safety_policy: Literal["enabled"] = "enabled"
    safety_policy_version: str
    agent_orchestration: Literal["enabled", "unavailable"]

File: apps/web/lib/api/system.ts
import { apiGet, hasString, isRecord } from "./client";
import type { SystemStatus } from "@/lib/types/api";

function isSystemStatus(value: unknown): value is SystemStatus {
  return (
    isRecord(value) &&
    hasString(value, "environment") &&
    hasString(value, "database") &&
    hasString(value, "upload_directory") &&
    value.vector_provider === "Pinecone" &&
    (value.vector_status === "connected" || value.vector_status === "unavailable") &&
    hasString(value, "vector_index") &&
    (value.vector_dimension === null || typeof value.vector_dimension === "number") &&
    (value.vector_metric === null || typeof value.vector_metric === "string") &&
    (value.vector_count === null || typeof value.vector_count === "number") &&
    value.embedding_provider === "Ollama" &&
    hasString(value, "embedding_model") &&
    (value.embedding_status === "connected" || value.embedding_status === "unavailable") &&
    (value.embedding_dimension === null || typeof value.embedding_dimension === "number") &&
    (value.compatibility === "compatible" || value.compatibility === "dimension_mismatch" || value.compatibility === "unavailable") &&
    value.llm_provider === "Ollama" &&
    hasString(value, "llm_model") &&
    (value.llm === "connected" || value.llm === "unavailable") &&
    value.safety_policy === "enabled" &&
    hasString(value, "safety_policy_version") &&
    (value.agent_orchestration === "enabled" || value.agent_orchestration === "unavailable")
  );
}

export function getSystemStatus(signal?: AbortSignal): Promise<SystemStatus> {
  return apiGet("/api/v1/system", isSystemStatus, signal, 15_000);
}
